src/resource/bindless.rs

```diff
@@ -0,0 +1,243 @@
+use std::sync::{Arc, Mutex};
+
+use anyhow::Result;
+
+use ash::vk;
+
+use crate::{pipeline::set_layout::DescriptorSetLayoutCreateInfo, util::cache::Resource};
+
+/// The maximum number of resources in a pool
+pub static MAX_BINDLESS_COUNT: u32 = 4096;
+
+/// A resource that can be used bindlessly.
+pub trait BindlessResource {
+    /// Get the [vk::DescriptorType] for this resource.
+    fn descriptor_type() -> vk::DescriptorType;
+
+    /// Get the [vk::DescriptorSetLayoutBinding] of this resource.
+    fn resource_binding(binding: u32) -> vk::DescriptorSetLayoutBinding {
+        vk::DescriptorSetLayoutBinding {
+            binding,
+            descriptor_type: Self::descriptor_type(),
+            descriptor_count: MAX_BINDLESS_COUNT,
+            stage_flags: vk::ShaderStageFlags::ALL,
+            p_immutable_samplers: std::ptr::null(),
+        }
+    }
+
+    /// Get the [vk::DescriptorImageInfo] of this resource.
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo;
+
+    /// Add this resource to `pool` and get a handle to it.
+    fn into_bindless(self, pool: &BindlessPool<Self>) -> BindlessHandle<Self>
+    where
+        Self: Sized
+    {
+        pool.alloc(self)
+    }
+}
+
+impl BindlessResource for crate::image::Image {
+    fn descriptor_type() -> vk::DescriptorType {
+        vk::DescriptorType::STORAGE_IMAGE
+    }
+
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo {
+        todo!()
+    }
+}
+
+/// Resource for a combined image sampler.
+#[derive(Debug)]
+pub struct CombinedImageSampler {
+    sampler: crate::sampler::Sampler,
```

Comment:
I just realized that we probably want to allow using the same sampler for multiple images. Maybe it makes sense to make this store an `Arc<Sampler>` or similar